### ISSUE TITLE
Automatically cleanup leaked JDBC connections

### ISSUE THREAD

**Issue by frowe** [OpenLiberty/open-liberty#10907]
UFO link: https://ibm.box.com/s/51kgxbr9dv16cra2x1cr3ye14m05p5dp
Traditional WAS detects if an unshared connection used in a transaction is not closed prior to the end of that transaction and closes it if autocommit = true, avoiding leaking of the connection.  Poorly written apps that don’t follow the best practice of get/use/close may rely on this behavior w/o being aware of it.  When those applications are migrated to Liberty, the lack of this cleanup may result in exhaustion of the connection pool.

**Comment by donbourne** [comment]
UFO review comments

```

communication
    - also add new release blog
    
feature design
    - add example of case where async requests may not want this behavior
    
migration 
    - should always enable this when migrating using the migration tools (if default is true then that means there's no work for migration tools)
    - for JDBC should always set this to "true" if we decide default for closeLeakedConnections should be "false"
    
serviceability
    - consider adding a info message when you have the closeLeakedConnections=false to warn people that they didn't close a connection when we think they should have
        - could use a warn / close / ignore semantic for closeLeadkedConnections (and would "warn" close it or not?)
        - warn case could include something in the log to show the stack trace of what path had leaked or even stack of where connection was checked out
        - include a stack of connection check out point when trace is enabled and connection is leaked or auto-closed
            - method in the security mnanager that will give you the call context more efficiently than a full stack walk - StackFinder
    
offline discussion needed
    - alasdair/nathan rauh/jim
        setting to true may still be better - alasdair not sold that the cases that require false are realistic
        - please let migration team (Dalia) know what decision is

put space in Open Liberty (not OpenLiberty)

config parm - closeLeakedConnections

end user overview - needs pictures for this - should be 

use case - prefer if we had a use case that described the value for any Liberty user, not in context of migration from tWAS

should add somewhere about this only applying when autoCommit is true (with regard to applicability to JCA)
    - alasdair believes autocommit is true for JMS in all cases - so we should potentially do this for JMS too

should mention that this only applies to LTCs and requires autoCommit to be true (and document this)

in 2 phase commit scenario, assume we won't close until commit is complete
```

**Comment by frowe** [comment]
Issue #11435 is resolved, we'll add an connectionManager attribute, name TBD, to control the cleanup behavior.  Default is false/off.  See design issue for discussion and rationale.

**Comment by njr-11** [comment]
@skasund here is a link to the STE. Please add the approval:
https://ibm.box.com/s/phiw1l2nr598f2m68k63acnzr96r9zte

**Comment by skasund** [comment]
@njr-11  Thanks for the STE slides.  I've approved the feature.

**Comment by njr-11** [comment]
@donbourne please add the focalApproved:serviceability label.

Serviceability Approval Comment:

WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?
_Yes, the whole point of this feature is to avoid customer issues by automatically cleaning up for them to avoid connection leaks, and to notify them with the stack so that they can also fix their application._

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
_Connection leaks in the application._
b) Who did you demo to?
_Greg Watts, Kyle Aure, Nathan Mittlestat_
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
_Yes._

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
_No one. It was determined that SVT was unnecessary due to exhaustive coverage in FAT, which we wrote ourselves._
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
_N/A._

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
_WAS: L2 CET
WAS: L3 JCA, Connection Management_

**Comment by chirp1** [comment]
Jim opened ID issue https://github.com/OpenLiberty/docs/issues/3691  However, because we have to bring over supporting information from WebSphere Liberty, we might not have the documentation done for 21003. We agreed that the blog post for this epic has the info that customers need to use the function until the documentation is written.
